### PR TITLE
[Merged by Bors] - refactor(Algebra/Category/ModuleCat): redefine presheaves of modules as families of objects in `ModuleCat`

### PR DESCRIPTION
**Pull request by joelriou**
In order to ease automation, a presheaf of modules is redefined as a family of objects in `ModuleCat`. Previously, it was defined as a presheaf of abelian groups with extra data, which was eventually creating a lot of issues with automation.

---

In #16755, this is applied to the construction of the free presheaf of modules functor (before the refactor, proofs were a nightmare, see #14245).

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Category/ModuleCat/Presheaf.lean [hunk 1]
```diff
@@ -33,148 +29,191 @@ universe v v₁ u₁ u
...
+    M.map f (r • m) = R.map f r • M.map f m)
 
-instance : Sub (P ⟶ Q) := ⟨fun f g => mk (f.hom - g.hom) (by
-  intros
-  rw [NatTrans.app_sub, AddMonoidHom.sub_apply, AddMonoidHom.sub_apply,
-    smul_sub, map_smul, map_smul])⟩
+/-- The object in `PresheafOfModules R` that is obtained from `M : Cᵒᵖ ⥤ Ab.{v}` such
+that for all `X : Cᵒᵖ`, `M.obj X` if a `R.obj X` module, in such a way that the
+restriction maps are semilinear. (This constructor should be used only in cases
+when the preferred constructor is `PresheafOfModules.mk` is not as convenient as this one.) -/
```
Comment (dagurtomas) [inline Mathlib/Algebra/Category/ModuleCat/Presheaf.lean:147]:
```suggestion
/-- The object in `PresheafOfModules R` that is obtained from `M : Cᵒᵖ ⥤ Ab.{v}` such
that for all `X : Cᵒᵖ`, `M.obj X` is a `R.obj X` module, in such a way that the
restriction maps are semilinear. (This constructor should be used only in cases
when the preferred constructor `PresheafOfModules.mk` is not as convenient as this one.) -/
```

**Review by dagurtomas — APPROVED** [review]
This is a very reasonable change. Everything looks good to me, I just had one very minor comment.

**Comment by joelriou** [comment]
Thanks @dagurtomas for the reviews!

**Comment by dagurtomas** [comment]
maintainer merge

**Comment by kim-em** [comment]
❤️  Very happy to see that this works out better!

bors merge

**Pull request closed without merging**